pretrain/openwebtext_trainer.py

```diff
@@ -0,0 +1,210 @@
+import math
+import sys
+import time
+from functools import partial
+from pathlib import Path
+from typing import Optional, Any
+
+import lightning as L
+import numpy as np
+import torch
+from lightning.pytorch.callbacks import ModelCheckpoint
+from lightning.pytorch.loggers import CSVLogger
+from lightning.pytorch.strategies import FSDPStrategy, XLAStrategy
+from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from lit_gpt import Config
+from lit_gpt.model import GPT, Block
+from lit_gpt.speed_monitor import measure_flops, estimate_flops, SpeedMonitorCallback
+from lit_gpt.utils import step_csv_logger, chunked_cross_entropy
+
+model_name = "pythia-70m"
+name = "openwebtext"
+out_dir = Path("out") / name
+data_dir = Path("data") / name
+save_interval = 1000
+eval_interval = 1000
+eval_iters = 100
+log_interval = 1
+
+# Hyperparameters
+learning_rate = 6e-4
+batch_size = 125
+micro_batch_size = 5
+gradient_accumulation_steps = batch_size // micro_batch_size
+assert gradient_accumulation_steps > 0
+max_iters = 600000  # num_epochs * (epoch_size // micro_batch_size) // devices
+weight_decay = 1e-1
+beta1 = 0.9
+beta2 = 0.95
+decay_lr = True
+warmup_iters = 2000
+lr_decay_iters = max_iters
+min_lr = 6e-5
+
+hparams = {k: v for k, v in locals().items() if isinstance(v, (int, float, str)) and not k.startswith("_")}
+
+
+class Trainer(L.Trainer):
+    # TODO: upstream this
+    def print(self, *args: Any, **kwargs: Any) -> None:
+        if self.local_rank == 0:
+            print(*args, **kwargs)
+
+
+class LightningGPTModule(L.LightningModule):
+    def __init__(self, trainer: Trainer, config: Config) -> None:
+        super().__init__()
+
+        with torch.device("meta"):
+            meta_model = GPT(config)
+            # estimated is too much of an optimistic estimate, left just for reference
+            estimated_flops = estimate_flops(meta_model) * micro_batch_size
+            trainer.print(f"Estimated TFLOPs: {estimated_flops * trainer.world_size / 1e12:.2f}")
+            x = torch.randint(0, 1, (micro_batch_size, config.block_size))
+            self.measured_flops = measure_flops(meta_model, x)
+            trainer.print(f"Measured TFLOPs: {self.measured_flops * trainer.world_size / 1e12:.2f}")
+            del meta_model, x
+
+        # TODO: hack until the Trainer supports `trainer.init_module()` or this can be done in `configure_sharded_model`
```

Comment:
We'll need to resolve this.

Reply:
Let's not do this here please 😅 
I suggest to instantiate the model normally. There won't be much of a difference with FSDP since `empty_init=False`. If you care about the precision, I suggest to do the

`torch.set_default_precision(...)` instead as a temporary workaround. 

The recommended pattern for Trainer currently is to put the creation of the model layers should go in `configure_sharded_model()` anyway (if using FSDP or DeepSpeed).

Reply:
This should stay for fast instantiation until it's upstreamed to the Trainer, it's ugly but it works and keeps us accountable for resolving it eventually